### PR TITLE
Add support for different encodings

### PR DESCRIPTION
**Pull request by Raudius**
* Resolves: #405 
  * #824 
  * #1633 
* Target version: master 

### Summary
Previously we would just assume files would use UTF-8. Now we make sure to convert files to UTF-8 when we open them. Without configuration just a small set of encodings are supported: 
```php
 'UTF-8', 'GB2312', 'GBK', 'BIG-5', 'SJIS-win', 'EUC-JP', 'Windows-1252', 'ISO-8859-15', 'ISO-8859-1', 'ASCII'
```

More encodings can be configured via the PHP config value: [`mbstring.detect_order`](https://www.php.net/manual/en/mbstring.configuration.php#ini.mbstring.detect-order)

### CODE REVIEW

File: lib/Service/EncodingService.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+namespace OCA\Text\Service;
+
+class EncodingService {
+	const COMMON_ENCODINGS = ['UTF-8', 'WINDOWS-1252', 'ISO-8859-15', 'ISO-8859-1', 'ASCII'];
+
+	public function encodeToUtf8(string $string): ?string {
```
Comment (vinicius73) [inline lib/Service/EncodingService.php:31]: It is possible to create a test for it @Raudius ?

Comment (Raudius) [reply lib/Service/EncodingService.php:31]: Yes, added! Nice suggestion :)

**Comment by vinicius73** [comment]
As we are using [mbstring](https://www.php.net/manual/pt_BR/book.mbstring.php) functions, it is important to define it in `composer.json`
Although it is already required by server.

https://github.com/nextcloud/server/blob/9e1d4d17ec792d8abcec2f9262472f5bb7a43acf/composer.json#L21-L29

```diff
--- a/composer.json
+++ b/composer.json
@@ -17,7 +17,8 @@
         }
     ],
     "require": {
-      "php": "^7.3|^8.0"
+      "php": "^7.3|^8.0",
+      "ext-mbstring": "*"
     },
     "scripts": {
```

What do you think?

**Review by vinicius73 — APPROVED** [review]

**Comment by juliushaertl** [comment]
@Raudius Could you do a rebase to update the branch rather then merge? That would make the commit history linear and a bit cleaner :)

**Comment by juliushaertl** [comment]
Found two more reportings on other encoding issues when doing a bit of triage:
https://github.com/nextcloud/text/issues/1633
https://github.com/nextcloud/text/issues/824

File: tests/unit/Service/EncodingServiceTest.php [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			$this->assertNotNull($detected_encoding);
+		}
+
+		$original_order = mb_detect_order();
+		$this->assertNotFalse(mb_detect_order($encoding));
+
+		$detected_encoding = $this->encodingService->detectEncoding(file_get_contents($file));
+		$this->assertEquals($encoding, $detected_encoding);
+
+		mb_detect_order($original_order);
```
Comment (juliushaertl) [inline tests/unit/Service/EncodingServiceTest.php:58]: Small nitpick on the code style 😉 I guess we currently don't run php-cs-fixer on the tests directory.
```suggestion
		$detectedEncoding = $this->encodingService->detectEncoding(file_get_contents($file));
		if ($this->isSupportedEncoding($encoding)) {
			$this->assertNotNull($detectedEncoding);
		}

		$originalOrder = mb_detect_order();
		$this->assertNotFalse(mb_detect_order($encoding));

		$detectedEncoding = $this->encodingService->detectEncoding(file_get_contents($file));
		$this->assertEquals($encoding, $detectedEncoding);

		mb_detect_order($originalOrder);
```

**Review by juliushaertl — APPROVED** [review]
Just a minor code style comment left on the tests, otherwise LGTM

**Pull request merged**